components/support/migration/src/main/java/mozilla/components/support/migration/FennecLoginsMigration.kt

```diff
@@ -0,0 +1,541 @@
+            from
+                moz_logins
+        """.trimIndent()
+
+        db.rawQuery(selectQuery, null).use {
+            while (it.moveToNext()) {
+                try {
+                    val encryptedUsername = it.getString(it.getColumnIndexOrThrow("encryptedUsername"))
+                    val encryptedPassword = it.getString(it.getColumnIndexOrThrow("encryptedPassword"))
+
+                    val derUsername = Base64.decode(encryptedUsername, Base64.DEFAULT)
+                    val plaintextUsername = derDecodeLoginEntry(derUsername)
+                        .decrypt(masterPassword, key4DbPath)
+                        .pkcs7unpad()
+
+                    val derPassword = Base64.decode(encryptedPassword, Base64.DEFAULT)
+                    val plaintextPassword = derDecodeLoginEntry(derPassword)
+                        .decrypt(masterPassword, key4DbPath)
+                        .pkcs7unpad()
+
+                    val encodedUsername = String(plaintextUsername, Charsets.UTF_8)
+                    val encodedPassword = String(plaintextPassword, Charsets.UTF_8)
+
+                    val hostname = it.getString(it.getColumnIndexOrThrow("hostname"))
+                    val httpRealm = it.getString(it.getColumnIndexOrThrow("httpRealm"))
+                    val formSubmitURL = it.getString(it.getColumnIndexOrThrow("formSubmitURL"))
+                    val usernameField = it.getString(it.getColumnIndexOrThrow("usernameField"))
+                    val passwordField = it.getString(it.getColumnIndexOrThrow("passwordField"))
+                    val guid = it.getString(it.getColumnIndexOrThrow("guid"))
+                    val timeCreated = it.getLong(it.getColumnIndexOrThrow("timeCreated"))
+                    val timesUsed = it.getInt(it.getColumnIndexOrThrow("timesUsed"))
+                    val timeLastUsed = it.getLong(it.getColumnIndexOrThrow("timeLastUsed"))
+                    val timePasswordChanged = it.getLong(it.getColumnIndexOrThrow("timePasswordChanged"))
+
+                    val fullRecord = ServerPassword(
+                        // 'id' is renamed to 'guid' as this record goes over the FFI boundary.
+                        id = guid,
+                        hostname = hostname,
+                        username = encodedUsername,
+                        password = encodedPassword,
+                        httpRealm = httpRealm,
+                        formSubmitURL = formSubmitURL,
+                        timesUsed = timesUsed,
+                        timeCreated = timeCreated,
+                        timeLastUsed = timeLastUsed,
+                        timePasswordChanged = timePasswordChanged,
+                        usernameField = usernameField,
+                        passwordField = passwordField
+                    )
+                    records.add(fullRecord)
+                } catch (e: Exception) {
+                    crashReporter.submitCaughtException(e)
```

Comment:
Here we'll send a caught exception to Sentry/Socorro for every incorrect record. How likely are incorrect records? Should we combine into one exception instead?

Reply:
Actually, yes, I think we should move this out of the loop. `decrypt` could throw a `NoSuchAlgorithmException` and then we'd be logging this a lot...